golangci-lint v2

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This updates Kubernetes to use the new and revised golangci-lint v2. On this occasion, the check suppressions got revisited, revealing that in some cases existing code really is broken (fix included in this PR, see individual commits).

#### Special notes for your reviewer:

This is a continuation of https://github.com/kubernetes/kubernetes/pull/131113. See last commit for the differences.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/cc @mmorel-35 @ldez

mmorel-35 committed:
chore: bump golangci-lint to v2

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

pohly committed:
DRA node test: fix useless gomega.Consistently

Passing a constant value to gomega.Consistently means that it will not re-check
while running.

Found by linter after removing the suppression rule for the check. It was
disabled earlier because of a bug in the linter.

pohly committed:
golangci-lint: finish migration to v2

This finishes the work started in
https://github.com/kubernetes/kubernetes/pull/131113.

Changes:
- Remove TODOs for things that we don't have plans to fix.
- Add issue for older TODO.
- Reorganize and remove suppression rules so that the base check
  has no unused rules.
- Document warn-unused, but don't enable it.
- Remove disabling of statistics (they are useful) and ensure that
  they don't get the ERROR prefix.
- Avoid ../ prefix in paths via `run.relative-path-mode: gomod`.

ldez commented on hack/golangci.yaml.in:
```diff
@@ -15,163 +15,120 @@ run:
 
 output:
-  sort-results: true
+  # Paths are relative to the hack/golangci-lint.yaml configuration:
+  #   ERROR: ../test/integration/apiserver/admissionwebhook/timeout_test.go:430:31: ...
+  #
+  # Joining with this prefix or any other non-empty path component makes the confusing ../ go away.
+  #   path-prefix: "hack"
+  #
```
```suggestion
```

I recommend removing this.

ldez commented on hack/golangci.yaml.in:
```diff
@@ -15,163 +15,120 @@ run:
+  # But then paths are still odd when the script is invoked inside a sub-directory
+  # because the path remains relative to the root of the repo. Absolute paths
+  # avoid the ambiguity.
+  path-mode: "abs"
```
You have 2 options:
- `run.relative-path-mode` used for path matching inside the configuration
- `output.path-mode` used to modify the output report paths

https://golangci-lint.run/usage/configuration/#run-configuration
https://golangci-lint.run/usage/configuration/#output-configuration

`run.relative-path-mode` defaults on `cfg` unless there is no configuration file, in this case, the `wd` mode is used.

pohly replied:
Changing `run.relative-path-mode` to `gomod` or `wd` had an unexpected side effect: the `settings.custom.logcheck.path: ../_output/local/bin/logcheck.so` no longer worked. It is relative to the configuration. Because the configuration is static, that setting cannot be an absolute path. I'm not sure whether `run.relative-path-mode` should affect lookup of plugins.

I probably can avoid that by using the new plugin mechanism that compiles logcheck into the golangci-lint binary.

pohly replied:
Why is `relative-path-mode: wd` not recommended (mentioned in the docs)? Any downsides that apply to Kubernetes?

pohly replied:
Let's shelf the idea of supporting invocation inside sub-directories for now. The script already didn't support that before (`cd "${KUBE_ROOT}"` !). I have support current working directory working locally, but let's consider that in a follow-up PR.

With `gomod` as `run.relative-path-mode` and a different lookup path it produces sane file names.

ldez replied:
> Why is `relative-path-mode: wd` not recommended (mentioned in the docs)?

Because the filepath inside the configuration will not work when running inside a package.

> Changing `run.relative-path-mode` to `gomod` or `wd` had an unexpected side effect: the `settings.custom.logcheck.path: ../_output/local/bin/logcheck.so` no longer worked

The `settings.custom.logcheck.path` is evaluated based on `relative-path-mode`.
So, if the path is not absolute, it will be relative to:
- the configuration file if the mode is `cfg`
- the `go.mod` file if the mode is `gomod`
- the launch location if the mode is `wd`

ldez replied:
The previous behavior (`wd`) was inconsistent because it was dependent on the location of the launch of golangci-lint.

The `relative-path-mode` controls all the path matching inside the configuration file (`path`, `path-expect`, etc.).

Depending on the mode, the paths inside the configuration should be relative to the same element.

I'm not sure to understand what you mean by "The main risk that I see then is that suppressions also don't work as expected"

pohly replied:
> I'm not sure to understand what you mean by "The main risk that I see then is that suppressions also don't work as expected"

If I use `wd` and invoke `hack/verify-golangci-lint.sh ./cmd/kubeadm/...` in the root, the `path-except: cmd/kubeadm` applies. If I invoke `hack/verify-golangci-lint.sh ./...` inside the `cmd/kubeadm` directory, the `path-except: cmd/kubeadm` does not match - at least that was my understanding and seems to be confirmed by your "The relative-path-mode controls all the path matching inside the configuration file (path, path-expect, etc.)." comment. I haven't actually tried it.

That is the behavior that I meant with "suppressions also don't work as expected". As a user, I would expect the behavior that I outlined earlier (paths in the config are always interpreted as relative to the location of the config, the current working directory doesn't matter for applying suppressions in the config).

ldez commented on hack/golangci.yaml.in:
```diff
@@ -185,84 +142,136 @@ linters:
+      checks:
+        - "all"
+        {{- if .Base }}
+        - "-QF1001"  # Apply De Morgan’s law
+        - "-QF1002"  # Convert untagged switch to tagged switch
+        - "-QF1003"  # Convert if/else-if chain to tagged switch
+        - "-QF1004"  # Use strings.ReplaceAll instead of strings.Replace with n == -1
+        - "-QF1006"  # Lift if+break into loop condition
+        - "-QF1007"  # Merge conditional assignment into variable declaration
+        - "-QF1008"  # Omit embedded fields from selector expression
+        - "-QF1009"  # Use time.Time.Equal instead of == operator
+        - "-QF1011"  # Omit redundant type from variable declaration
+        - "-QF1012"  # Use fmt.Fprintf(x, ...) instead of x.Write(fmt.Sprintf(...))
```
```suggestion
        - "-QF*"
```

ldez commented on hack/tools/go.mod:
```diff
@@ -8,68 +8,77 @@ require (
 	github.com/cespare/prettybench v0.0.0-20150116022406-03b8cfe5406c
 	github.com/client9/misspell v0.3.4
-	github.com/golangci/golangci-lint v1.64.5
+	github.com/golangci/golangci-lint/v2 v2.1.2
```
```suggestion
	github.com/golangci/golangci-lint/v2 v2.1.5
```

pohly replied:
... which is now also on https://golangci-lint.run/product/changelog. I looked earlier today and unless I missed it, it wasn't mentioned yet.

ldez replied:
It was in the changelog: the changelog is published before the release.
I think it's just a browser cache effect.

pohly commented on hack/golangci.yaml.in:
```diff
@@ -14,164 +14,115 @@
-      text: should not be used because managedFields was removed
-      path: _test.go$
+    # Log a warning if an exclusion rule is unused.
+    warn-unused: true
```
Not so nice about this one:

```console
$ hack/verify-golangci-lint.sh ./staging/src/k8s.io/dynamic-resource-allocation
...
running /nvme/gopath/src/k8s.io/kubernetes/_output/local/bin/golangci-lint run --color=always --config=/nvme/gopath/src/k8s.io/kubernetes/hack/golangci.yaml ./staging/src/k8s.io/dynamic-resource-allocation
ERROR: level=warning msg="[runner/exclusion_paths] The pattern \"third_party\" match no issues"
ERROR: level=warning msg="[runner/exclusion_rules] Skipped 0 issues by rules: [Text: \"lostcancel|printf\", Linters: \"govet\"]"
ERROR: level=warning msg="[runner/exclusion_rules] Skipped 0 issues by rules: [Text: \"wrong error assertion. Consider using `gomega.(Eventually|Consistently)|wrong comparison assertion|wrong length assertion\", Linters: \"ginkgolinter\"]"
ERROR: level=warning msg="[runner/exclusion_rules] Skipped 0 issues by rules: [Text: \"append result not assigned to the same slice|put a space between `//` and comment text|sloppyLen|elseif|should rewrite switch statement to if statement|regexpMust|wrapperFunc: use strings.ReplaceAll|singleCaseSwitch|deprecatedComment|exitAfterDefer|captLocal|unlambda|underef|unslice|valSwap|typeSwitchVar\", Linters: \"gocritic\"]"
ERROR: level=warning msg="[runner/exclusion_rules] Skipped 0 issues by rules: [Text: \"should not use Add, use AddVersioned instead\", Path: \"_test.go$\", Linters: \"forbidigo\"]"
ERROR: level=warning msg="[runner/exclusion_rules] Skipped 0 issues by rules: [Text: \"assignOp:\", Linters: \"gocritic\"]"
ERROR: level=warning msg="[runner/exclusion_rules] Skipped 0 issues by rules: [Text: \"ifElseChain: rewrite if-else to switch statement\", Linters: \"gocritic\"]"
ERROR: level=warning msg="[runner/exclusion_rules] Skipped 0 issues by rules: [Text: \"comment on exported (method|function|type|const)|should have( a package)? comment|comment should be of the form|comment on exported (method|function|type|const)|should have( a package)? comment|comment should be of the form|exported (.+) should have comment( \\\\(or a comment on this block\\\\))? or be unexported|package comment should be of the form \\\"(.+)...|comment on exported (.+) should be of the form \\\"(.+)...|should have a package comment\", Path Except: \"cmd/kubeadm\", Linters: \"revive, staticcheck\"]"
ERROR: level=warning msg="[runner/exclusion_rules] Skipped 0 issues by rules: [Path: \"staging/src/k8s.io/client-go/util/workqueue/metrics.go\", Linters: \"unused\"]"
ERROR: level=warning msg="[runner/exclusion_rules] Skipped 0 issues by rules: [Text: \"should not be used because managedFields was removed\", Path: \"_test.go$\", Linters: \"forbidigo\"]"
ERROR: level=warning msg="[runner/exclusion_rules] Skipped 0 issues by rules: [Path: \"(pkg/volume/*|test/*|azure/*|pkg/cmd/wait*|request/bearertoken/*|metrics/*|filters/*)\", Linters: \"gocritic\"]"
ERROR: 0 issues.
Congratulations! All files are passing lint :-)
```

I don't get the non-zero exit code that @mmorel-35 mentioned, but the `ERROR` log output is confusing. I'm leaning towards commenting out this setting.

@ldez: would it make sense to add a command line flag for it? Then our CI job could use that flag when checking the entire code base. In that case, a non-zero exit code would be useful because the `ERROR` log output will be overlooked when it gets emitted in a successful job.

ldez replied:
> would it make sense to add a command line flag for it?

golangci-int v2 has two exclusion sections, so adding a flag only for one doesn't make sense.

Also, I don't know where the log format you have comes from because this is not the golangci-lint format:
```console
$ golangci-lint run
WARN [runner/exclusion_rules] Skipped 0 issues by rules: [Path: "test/testdata_etc", Linters: "gofmt, goimports"] 
WARN [runner/exclusion_rules] Skipped 0 issues by rules: [Path: "internal/go", Linters: "gofmt, goimports"] 
WARN [runner/exclusion_rules] Skipped 0 issues by rules: [Path: "pkg/goformatters/gci/internal", Linters: "gofmt, goimports"] 
WARN [runner/exclusion_rules] Skipped 0 issues by rules: [Path: "pkg/commands/internal/migrate/two/", Linters: "lll"] 
WARN [runner/exclusion_rules] Skipped 0 issues by rules: [Path: "internal/x", Linters: "gofmt, goimports"] 
WARN [runner/exclusion_rules] Skipped 0 issues by rules: [Path: "pkg/goanalysis/runner_checker.go", Linters: "gofmt, goimports"] 
0 issues.
```

pohly replied:
True, "ERROR" is a prefix added by `golangci-lint.sh` to make the output stand out more in the Kubernetes job viewer. The expectation (true before) was that any output from golangci-lint is related to an issue.

This no longer true now (these warnings, the stats at the end), so I need to make this better.

pohly replied:
Any idea why I was getting `level=warning` and you got `WARN` instead?

ldez replied:
This also makes me wonder.
:thinking: maybe it's related to a logrus behavior (maybe an env var or something like that)

pohly replied:
logrus is v1.9.3, as in the golangci-lint module. I'm not aware of any particular env variable either (we don't normally use logrus in Kubernetes). :shrug:

ldez replied:
This is because a TTY is not attached: https://github.com/sirupsen/logrus

ldez commented:
A bit off topic, I see you are still using [`github.com/client9/misspell`](https://github.com/client9/misspell), this project has been archived.
I recommend using the fork we maintain: [`github.com/golangci/misspell`](https://github.com/golangci/misspell)

pohly commented:
@mmorel-35: can you perhaps review and LGTM? You have most context on this.

pohly commented:
/assign @BenTheElder 

For approval and another review. It already has informal LGTM from @mmorel-35.

BenTheElder commented on hack/golangci.yaml:
```diff
@@ -14,155 +14,109 @@
 
+  # The default is relative to the configuration, which is confusing because
+  # then all paths start with ../ to move out of the "hack" directory.
+  # `gomod` mirrors the current behavior of `golangci-lint.sh` changing into
```
This makes sense to me, this is essentially the repo root, at least for the main module.

That will make other paths more predictable anyhow.

BenTheElder commented on hack/golangci.yaml:
```diff
@@ -173,127 +127,181 @@ linters:
+    staticcheck:
+      checks:
+        - "all"
+        - "-QF1001"  # Apply De Morgan’s law
```
I guess we can incrementally identify some of these to eliminate, many sound very opinionated, some of them sound like possible bugs.

BenTheElder commented:
/triage accepted

BenTheElder commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion.go:
```diff
@@ -65,7 +65,7 @@ func Convert_v1_JSON_To_apiextensions_JSON(in *JSON, out *apiextensions.JSON, s
 		*out = i
 	} else {
-		out = nil
+		*out = nil
```
this is a really subtle one, we'll need to tag in API reviewers

pohly replied:
Without this:

```
ERROR: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion.go:68:3: ineffectual assignment to out (ineffassign)
ERROR: 		out = nil
ERROR: 		^
```

BenTheElder commented on hack/tools/go.mod:
```diff
@@ -7,69 +7,77 @@ godebug default=go1.24
 	github.com/aojea/sloppy-netparser v0.0.0-20210819225411-1b3bd8b3b975
 	github.com/cespare/prettybench v0.0.0-20150116022406-03b8cfe5406c
-	github.com/golangci/golangci-lint v1.64.5
+	github.com/golangci/golangci-lint/v2 v2.1.5
```
starting to think about moving tools into seperate modules, so they behave closer to if you just `go install`ed them and don't conflate dep versions, there are some tradeoffs (sharing deps improves build caching for them marginally), it's hard to tell if any of these impact other tools

ldez replied:
if you want to do that, I recommend reading the section `go tool` of the golangci-lint doc https://golangci-lint.run/welcome/install/#install-from-sources

BenTheElder replied:
Right, currently we have this issue from the other tools:
> When using the "tools pattern" or tool command/directives, the dependencies of a tool can modify the dependencies of another tool or your project. The resulting binary was not tested and is not guaranteed to work.

Tracking it in a shared tools module has risks, I think we should split it out.
Then the only question is if we use a specific module or avoid a module entirely.
Using a module means we get GOSHASUM and other visibility, at the risk of breaking replace statements.

BenTheElder replied:
We prefer not to take opaque binary dependencies generally, and it's also more complex to do portably across OS-es. It is somewhat less opaque to track this way, and I see shasums breaking *as a feature* that indicates some unknown change was made that hasn't been looked at. The GOSHASUMS enable TOFU in a way that is much simpler and re-usable than binaries.

ldez replied:
I suggested that you read the section "`go tool` usage recommendations`.
You have to click on the triangle.

BenTheElder replied:
Ahh! it wasn't obvious that this expanded, and the lines above at the linked heading talk about `go tool`

Might be worth tweaking the styling / layout.

ldez replied:
It is intentionally not a heading because we strongly don't recommend that.
But I prefer to provide the right way to do it instead of letting users in the dark.
So this is a compromise.

pohly replied:
Installing golangci-lint through a separate module seems worthwhile. I've added one more commit doing that:

```
golangci-lint: move into hack/tools/golangci-lint

The advantage is that it separates the dependencies of the different tools
better. golangci-lint in particular has many dependencies and is sometimes
sensitive to the exact version being used. This way, "go get" bumps up
dependencies exactly as defined by the upstream golangci-lint module.

It's not quite self-contained because logcheck as a Go plugin for golangci-lint
must be built from the same dependencies. But it only adds one and does not
change any of the others.

While at it, the Go 1.24 "tools" directive gets used instead of the traditional
tools.go approach.
```

ldez replied:
As explained inside the doc, you don't need a "full" new module, you can use only a dedicated module file.
I don't say this is a better approach, this is just to be sure this approach has been evaluated.

BenTheElder reviewed (commented):
/approve
for hack/
Need more eyes on the other changes to confirm.

pohly committed:
golangci-lint: move into hack/tools/golangci-lint

The advantage is that it separates the dependencies of the different tools
better. golangci-lint in particular has many dependencies and is sometimes
sensitive to the exact version being used. This way, "go get" bumps up
dependencies exactly as defined by the upstream golangci-lint module.

It's not quite self-contained because logcheck as a Go plugin for golangci-lint
must be built from the same dependencies. But it only adds one and does not
change any of the others.

While at it, the Go 1.24 "tools" directive gets used instead of the traditional
tools.go approach.

liggitt commented on pkg/apis/autoscaling/v2beta2/conversion.go:
```diff
@@ -48,7 +48,6 @@ func Convert_autoscaling_HorizontalPodAutoscaler_To_v2beta2_HorizontalPodAutosca
 	}
 	if behavior.ScaleUp != nil && behavior.ScaleUp.Tolerance != nil {
 		if !copiedAnnotations {
-			copiedAnnotations = true
```
I want to keep this, even if we just have to ignore the lint... dropping this means that the variable is incorrect after this point if a later check / change was added

pohly replied:
Changed.

liggitt commented on hack/golangci.yaml.in:
```diff
@@ -185,84 +147,136 @@ linters:
+      checks:
+        - "all"
+        {{- if .Base }}
+        - "-QF1001"  # Apply De Morgan’s law
```
are all these exclusions for checks that are new in v2, or did we start running entire linters we weren't running before? I'm surprised we have to exclude so many individual new checks

liggitt commented on hack/golangci.yaml.in:
```diff
@@ -14,164 +14,126 @@
+  # The default is relative to the configuration, which is confusing because
+  # then all paths start with ../ to move out of the "hack" directory.
+  # `gomod` mirrors the current behavior of `golangci-lint.sh` changing into
+  # the root.
```
> ... changing into the root

the root of `k8s.io/kubernetes` or of the module being evaluated (when we evaluate stuff under `staging/src/k8s.io/*`?

pohly replied:
Changed:

```
# `gomod` mirrors the current behavior of `golangci-lint.sh` changing into
# the root of the repository. Because we are operating in a workspace,
# the module picked by `gomod` is the main one.
```

liggitt commented on hack/golangci.yaml.in:
It's tricky to understand how stuff in the old config translated to the new one. Was the v2 config auto-generated with https://golangci-lint.run/product/migration-guide/ or done by hand?

Is there a way to see which changes were mechanical v1 → v2 translations and which were intentional changes from how we configured v1 to behave?

pohly replied:
I think @mmorel-35 started by converting automatically because the initial commit in https://github.com/kubernetes/kubernetes/pull/131113 also included some re-ordering. 

But some things are necessarily manual. I think we tried to stick as closely as possible to the original configuration, except for a few points where some cleanup was possible because something was obsolete (for example, a suppression because of a bug in the Ginkgo linter which had been fixed already long ago).

The important part is that the current code passes the base configuration. Not much had to be changed in the code to achieve that. If we find that the new configuration is too strong (whether its the base configuration or for hints), then we can always refine it in the future.

ldez replied:
The re-ordering is not related to automatic conversion, because the order is not the order produced by the automatic conversion.

pohly replied:
Not anymore, because at some point I asked @mmorel-35 to restore the original order.

@mmorel-35: did you initially convert automatically?

ldez commented on hack/tools/golangci-lint/go.mod:
```diff
@@ -0,0 +1,211 @@
+
+tool (
+	github.com/golangci/golangci-lint/v2/cmd/golangci-lint
+	sigs.k8s.io/logtools/logcheck
```
The `github.com/golangci/plugin-module-register/register` is working with v2, there are no changes on that.

haircommander commented:
/cc

pohly commented on pkg/controller/disruption/disruption.go:
```diff
@@ -644,7 +646,7 @@ func (dc *DisruptionController) processNextWorkItem(ctx context.Context) bool {
 	}
 
-	utilruntime.HandleError(fmt.Errorf("Error syncing PodDisruptionBudget %v, requeuing: %w", dKey, err)) //nolint:stylecheck
+	utilruntime.HandleErrorWithContext(ctx, err, "Error syncing PodDisruptionBudget", "key", dKey)
```
> was the existing //nolint:stylecheck comment suppressing an "errors start with lowercase letter" check?

That check is not enforced. Perhaps `//nolint` was added in https://github.com/kubernetes/kubernetes/blame/03a3c0c89161935bc2338f5754ebb1104f779af1/pkg/controller/disruption/disruption.go#L647 to silence the hint. That was unnecessary and in this case, I would have preferred fixing the code.

We should have a linter which complains about `//nolint:something` without an additional comment after it (`//nolint: // " Not relevant because ...`)...

> If so, this changed to context logging (which is a functional change that looks unrelated to the linting), 

We can:
- Leave this line as it is (including the `//nolint:stylecheck`, which is useless now because the linter is no longer active).
- Remove just the `//nolint:stylecheck`.
- Update the line to use current best practices (= use `HandleErrorWithContext`).

I prefer the last option.

> but is still passing the initial uppercase message the stylecheck complained about?

`HandleError` was a logging function in disguise with additional tooling that normal error logging doesn't have. Using `fmt.Errorf` was (mis)used to produce a log message.

`HandleErrorWithContext` explicitly accepts a log message and additional key/value pairs, in addition to the error (if there is one - it's not required, i.e. no need to fabricate a new one if there isn't one already). For log messages, the recommendation is to start with a capital letter. I'm not exactly sure why, that recommendation pre-dates my involvement.

TLDR: `Error syncing" is fine here.

> At the very least, we should avoid dropping the "requeuing" bit from the message, so "Error syncing PodDisruptionBudget, requeuing"

Ack, changed.

ldez replied:
This problem is not a specific issue related to nolinlint, this is explained inside the issue.

pohly commented on hack/verify-golangci-lint.sh:
```diff
@@ -153,9 +155,8 @@ if [ "${golangci_config}" ]; then
   # don't to do it when not required.
   if grep -q 'path: ../_output/local/bin/' "${golangci_config}" &&
```
Yes, thanks for catching this.

pohly commented on hack/verify-golangci-lint.sh:
```diff
@@ -121,14 +121,22 @@ done
+# pre-built binaries for different platforms and gives more insights on dependencies.
 echo "installing golangci-lint and logcheck plugin from hack/tools into ${GOBIN}"
-GOTOOLCHAIN="$(kube::golang::hack_tools_gotoolchain)" go -C "${KUBE_ROOT}/hack/tools" install github.com/golangci/golangci-lint/v2/cmd/golangci-lint
+GOWORK=off GOTOOLCHAIN="$(kube::golang::hack_tools_gotoolchain)" go -C "${KUBE_ROOT}/hack/tools/golangci-lint" install github.com/golangci/golangci-lint/v2/cmd/golangci-lint
```
That also works. Changed.

pohly commented:
/test pull-kubernetes-integration

TestVolumeBindingRescheduling timeout
